Std/Data/List/Count.lean

```diff
@@ -0,0 +1,261 @@
+      · rfl
+      · simp only [h]
+
+theorem countp_eq_length_filter (l) : countp p l = length (filter p l) := by
+  induction l with
+  | nil => rfl
+  | cons x l ih =>
+    by_cases h : p x
+    · rw [countp_cons_of_pos p l h, ih, filter_cons_of_pos l h, length]
+    · rw [countp_cons_of_neg p l h, ih, filter_cons_of_neg l h]
+
+theorem countp_le_length : countp p l ≤ l.length := by
+  simp only [countp_eq_length_filter]
+  apply length_filter_le
+
+@[simp] theorem countp_append (l₁ l₂) : countp p (l₁ ++ l₂) = countp p l₁ + countp p l₂ := by
+  simp only [countp_eq_length_filter, filter_append, length_append]
+
+theorem countp_pos : 0 < countp p l ↔ ∃ a ∈ l, p a := by
+  simp only [countp_eq_length_filter, length_pos_iff_exists_mem, mem_filter, exists_prop]
+
+@[simp] theorem countp_eq_zero : countp p l = 0 ↔ ∀ a ∈ l, ¬p a := by
+  simp only [countp_eq_length_filter, length_eq_zero, filter_eq_nil]
+
+@[simp] theorem countp_eq_length : countp p l = l.length ↔ ∀ a ∈ l, p a := by
+  rw [countp_eq_length_filter, filter_length_eq_length]
+
+theorem Sublist.countp_le (s : l₁ <+ l₂) : countp p l₁ ≤ countp p l₂ := by
+  simp only [countp_eq_length_filter]
+  apply s.filter.length_le
+
+@[simp] theorem countp_filter (l : List α) :
+    countp p (filter q l) = countp (fun a => p a ∧ q a) l := by
+  simp only [countp_eq_length_filter, filter_filter]
+
+@[simp] theorem countp_true {l : List α} : (l.countp fun _ => true) = l.length := by simp
+
+@[simp] theorem countp_false {l : List α} : (l.countp fun _ => false) = 0 := by simp
+
+@[simp] theorem countp_map (p : β → Bool) (f : α → β) :
+    ∀ l, countp p (map f l) = countp (p ∘ f) l
+  | [] => rfl
+  | a :: l => by rw [map_cons, countp_cons, countp_cons, countp_map p f l]; rfl
+
+variable {p q}
+
+theorem countp_mono_left (h : ∀ x ∈ l, p x → q x) : countp p l ≤ countp q l := by
+  induction l with
+  | nil => apply Nat.le_refl
+  | cons a l ihl =>
+    rw [forall_mem_cons] at h
+    have ⟨ha, hl⟩ := h
+    simp [countp_cons]
+    cases h : p a
+    . simp
+      apply Nat.le_trans ?_ (Nat.le_add_right _ _)
+      apply ihl hl
+    . simp [ha h, Nat.add_one]
+      apply Nat.succ_le_succ
+      apply ihl hl
+
+theorem countp_congr (h : ∀ x ∈ l, p x ↔ q x) : countp p l = countp q l :=
+  Nat.le_antisymm
+    (countp_mono_left fun x hx => (h x hx).1)
+    (countp_mono_left fun x hx => (h x hx).2)
+
+end Countp
+
+/-! ### count -/
+
+section Count
+
+variable [DecidableEq α]
+
+@[simp] theorem count_nil (a : α) : count a [] = 0 := rfl
+
+theorem count_cons (a b : α) (l : List α) :
+    count a (b :: l) = count a l + if a = b then 1 else 0 := by conv =>
+  simp [count, countp_cons]
+  lhs
+  simp only [eq_comm]
+
+@[simp] theorem count_cons_self (a : α) (l : List α) : count a (a :: l) = count a l + 1 := by
+  simp [count_cons]
+
+@[simp] theorem count_cons_of_ne (h : a ≠ b) (l : List α) : count a (b :: l) = count a l := by
```

Comment:
I think (?) we could just mark `count_cons` simp, right?

Reply:
I don't think we should be putting an expression like that in the goal if people don't specifically ask for it.